edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,185 @@
+                child_process.shutdown_if_running();
+            }
+
+            has_shutdown_listener.store(true, Ordering::Relaxed);
+        });
+
+        has_shutdown
+    }
+
+    fn shutdown_if_running(&mut self) {
+        if self.is_running() {
+            info!("Terminating {:?} process", self.name);
+            self.send_signal(Signal::SIGTERM);
+        } else {
+            info!("{:?} process has stopped", self.name)
+        }
+
+        self.wait_for_exit();
+
+        if self.is_running() {
+            info!("Killing {:?} process", self.name);
+            self.send_signal(Signal::SIGKILL);
+        }
+    }
+
+    // TODO: get signal name and log it
+    fn send_signal(&mut self, signal: Signal) {
+        if let Err(e) = signal::kill(Pid::from_raw(self.process.id() as i32), signal) {
+            error!("Failed to send signal to {:?} process. {:?}", self.name, e);
+        }
+    }
+
+    fn is_running(&mut self) -> bool {
+        match self.process.try_wait() {
+            Ok(status) => status.is_none(),
+            Err(e) => {
+                error!(
+                    "Error while polling {:?} process status. {:?}",
+                    self.name, e
+                );
+                false
+            }
+        }
+    }
+
+    fn wait_for_exit(&mut self) {
+        let mut elapsed_secs = 0;
+        while elapsed_secs < PROCESS_SHUTDOWN_TOLERANCE_SECS && self.is_running() {
+            let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+            thread::sleep(poll_interval);
+            elapsed_secs += PROCESS_POLL_INTERVAL_SECS;
+        }
+    }
+}
+
+fn main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_shutdown_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!(
+                "Failed to register sigterm listener. Shutting down. {:?}",
+                e
+            );
+            exit(1);
+        }
+    };
+
+    let has_broker_shutdown = ChildProcess::run_child_process(
+        "MQTT Broker".to_string(),
+        Command::new("/usr/local/bin/mqttd").stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    let has_edgehub_shutdown = ChildProcess::run_child_process(
+        "Edge Hub".to_string(),
+        Command::new("dotnet")
+            .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+            .stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    while !should_shutdown.load(Ordering::Relaxed) {
+        if has_broker_shutdown.load(Ordering::Relaxed) {
+            break;
+        }
+
+        if has_edgehub_shutdown.load(Ordering::Relaxed) {
+            break;
+        }
+
+        thread::sleep(Duration::from_secs(1));
+    }
+
+    should_shutdown.store(true, Ordering::Relaxed); // tell the threads to shut down their child process
```

Comment:
We use a shared bool to signal the other threads to shut down the child processes. The threads also share bool that indicates their running/stopped status, so we can wait on this before returning from main.